### ISSUE TITLE
Enable consuming of ES modules in extensions

### ISSUE THREAD

**Issue by igorskyflyer** [microsoft/vscode#130367]
When developing extensions and using either JavaScript or TypeScript we are unable to consume ES modules, only somewhat legacy CommonJS modules, setting the `type` to `module` and rewriting the extension to use `import` instead of `require` breaks the extension, generating an exception that states that all modules should use `import` instead of `require` in internal VS Code JavaScript files, I conclude it's caused by the `type: module` that forces Node to treat all `.js` files as ES modules. Tried using TypeScript which transpiles its own syntax to CommonJS module - so that's a no, I have also tried using just `.mjs` extension, again the same issue.

What is the status of this issue and are there plans to enable using of ES modules in extension development? That (could) bring somewhat big performance gains when bundling extensions with, for example, [`esbuild`](https://github.com/evanw/esbuild) because it would enable tree-shaking - dead code removal, thus loading only necessary code. But I think this is not an extension API only issue, right? This needs to be done for VS Code itself?

**Comment by igorskyflyer** [comment]
Closing after 15 days of no reply.

**Comment by alexdima** [comment]
cc @jrieken

**Comment by Lemmingh** [comment]
Sounds like a duplicate of #116056. Is it possible to reopen that issue?

---

The VS Code extension host currently only accepts CJS module, as shown in the (trimmed) error message below:

```log
Activating extension failed

Error [ERR_REQUIRE_ESM]: Must use import to load ES Module
require() of ES modules is not supported.

Instead change the requiring code to use import()

at internal/modules/cjs/loader.js:823:14

at require (internal/modules/cjs/helpers.js:88:18)
at Function.t [as __$__nodeRequire] (c:\Program Files\Microsoft VS Code\resources\app\out\vs\loader.js:5:101)
at v._loadCommonJSModule
at v._doActivateExtension
at v._activateExtension (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:99:11695)
```

However, after reading the [Web Extensions](https://code.visualstudio.com/api/extension-guides/web-extensions) authoring guide, I think ES module will be more convenient in future.

Besides, TypeScript can transpile [dynamic `import()` in CJS module](https://nodejs.org/docs/latest-v14.x/api/esm.html#esm_import_statements) in a surprising way unless enabling `esModuleInterop`, while it's never a problem in ES module where `import()` is emitted as is. VS Code extension authors can benefit from building extension as ES module.

**Comment by andschwa** [comment]
For what it's worth, I ran into this trying to update `node-fetch` from 2.x to 3.x, since the newer version is ESM only. Looks to be the direction things are going, so extension developers are only more likely in the future to run into issues due to lack of ESM support.

**Comment by andschwa** [comment]
FYI @TylerLeonhardt one of the Code issues I'd love to see fixed 😃

**Comment by andschwa** [comment]
Excuse me, @TylerLeonhardt, do you know why my comment above was marked as spam? I have to assume that was a mistake.

**Comment by alexdima** [comment]
@andschwa I'm sorry, it might have been me that marked the comment as spam, but I don't remember doing it. I personally tend to hide comments that do not bring any value to the underlying discussion or distract from it.

**Comment by jasonwilliams** [comment]
> TypeScript 4.5 will _perhaps_ have a new `module` option called `node12`, which preserves `import()` in CJS module. Then, although your entry point still have to be a CJS module now, you can load ES modules internally in an **asynchronous** manner.
> 
> See
> 
> * https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/
> * [TypeScript 4.5 Iteration Plan TypeScript#45418 (comment)](https://github.com/microsoft/TypeScript/issues/45418#issuecomment-953399609)
> * The links in [my comment](https://github.com/microsoft/vscode/issues/130367#issuecomment-912714384)

This has now been done in https://github.com/microsoft/TypeScript/pull/45884

**Comment by Silic0nS0ldier** [comment]
Conversation here seems to have drifted off topic. While `node12` module resolution logic in TypeScript will help with extension authoring (and directly consuming ES module packages in VSCode), it is not required to support ES modules. Extensions are regular JavaScript, so all that needs to happen is for dynamic import support to be introduced.

I aim to author ESM wherever I can (easier to share code across the ecosystem) so I had a look at the VSCode source to see if I can push this forward.

Best I can tell, there are 2 environments we care about (within the scope of this issue, web extensions have additional limitations so I'm ignoring it for now). They can run in the main thread (which is hopefully rare) or in the extension host.

#### Extension Host

What i've managed to gleam from VSCode source is that the extension host is started with the `worker_threads` API, which potentially could be imposing a dynamic require restriction (I'll test this in NodeJS to clarify).

https://github.com/microsoft/vscode/blob/4a1a8b07ff093e7dcb8067c185eec49502b56aa5/src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts#L12
https://github.com/microsoft/vscode/blob/4a1a8b07ff093e7dcb8067c185eec49502b56aa5/src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts#L22-L24

Extensions hosted here are imported with `require` (or rather what I believe is a webpack escaped version). I've included more of the call chain in the snippets below to help with context.

https://github.com/microsoft/vscode/blob/9fd6ee7095e76a304556c06bbef16fc4f05cf697/src/vs/workbench/api/common/extHostExtensionService.ts#L171-L172

https://github.com/microsoft/vscode/blob/9fd6ee7095e76a304556c06bbef16fc4f05cf697/src/vs/workbench/api/common/extHostExtensionService.ts#L394

https://github.com/microsoft/vscode/blob/9fd6ee7095e76a304556c06bbef16fc4f05cf697/src/vs/workbench/api/node/extHostExtensionService.ts#L103

#### Main Thread

This follows a different path, but appears to enter the extension host before being redirected back to the main thread (likely necessary to orchestrate instantiation of extension dependencies).

https://github.com/microsoft/vscode/blob/9fd6ee7095e76a304556c06bbef16fc4f05cf697/src/vs/workbench/api/common/extHostExtensionService.ts#L167-L170

https://github.com/microsoft/vscode/blob/788e39aad864837eb3b74e3d935d52594f4c8c96/src/vs/workbench/api/browser/mainThreadExtensionService.ts#L48-L50

https://github.com/microsoft/vscode/blob/e24174932025cb302c706d62b6dcae76d010fe83/src/vs/workbench/services/extensions/common/abstractExtensionService.ts#L955

And that is about as far as I could get.

**Comment by Lemmingh** [comment]
AFAIK, generally speaking, VS Code extension hosts and VS Code extensions are complied by TypeScript and run on the same engines, thus, loading ES modules in VS Code extension hosts and VS Code extensions is essentially the same problem.

---

Let's take a look at the **runtime** first.

There're two extension hosts. One is on Node.js. The other is on web worker.

Node.js's support for ES module is usable, although some features are experimental.

> ## Module loading methods in Node.js
>
> | Method    | L: ES | L: CJS | A: ES | A: CJS |
> | --------- | ----- | ------ | ----- | ------ |
> | import    | Y     | Y      | Y     | -      |
> | import()  | Y     | Y      | Y     | Y      |
> | require() | -     | Y      | -     | Y      |
>
> * ES: ECMAScript module
> * CJS: CommonJS module
> * L: Can Load
> * A: Available In
> * Y: Yes
> * -: No

Web browsers' support is a bit awkward, as ES module is still not available in web workers on [Firefox](https://bugzil.la/1247687).

> <https://caniuse.com/mdn-javascript_statements_import_worker_support>

---

TypeScript, as part of the **toolchain**, is critical. But when setting `"module": "CommonJS"`, the dynamic `import()` expressions are replaced with `require()` calls, which is fatal. That's why people pay great attention to changes in TypeScript.

Bundlers, such as webpack and Rollup, also have a few problems with ES module.

---

We can learn from #135450 that VS Code needs to intercept the loading of `vscode` to control **extension API** access. However, I do not think that's a problem, because we can configure bundlers to always emit `require("vscode")`, so that the extension hosts won't have to worry about other module systems.

Here is a **very simple** example using webpack.

<details>
<summary>Source: main.ts</summary>

```typescript
import * as vscode from "vscode";
import { hello } from "./sub";

export function activate(context: vscode.ExtensionContext) {
    vscode.window.showInformationMessage(hello);
}
```

</details>
<details>
<summary>Source: sub.ts</summary>

```typescript
export const hello = "Hello.";
export const nice = "Nice!";
```

</details>
<details>
<summary>tsconfig.json</summary>

```json
{
    "compilerOptions": {
        "module": "ES2020",
        "target": "ES2020",
        "lib": ["ES2019", "WebWorker"],
        "moduleResolution": "Node",
        "strict": true
    },
    "files": ["main.ts", "sub.ts"]
}
```

</details>
<details open>
<summary>webpack.config.js</summary>

```javascript
// @ts-check
"use strict";
const path = require("path");

/** @typedef {import("webpack").Configuration} WebpackConfig **/

/** @type {WebpackConfig} */
const Config_Base = {
    mode: "none",
    entry: { main: "./main.ts" },
    resolve: { extensions: [".ts"] },
    module: {
        rules: [
            {
                test: /\.ts$/,
                exclude: /node_modules/,
                use: [{ loader: "ts-loader" }],
            },
        ],
    },

    // The ES module output is experimental.
    experiments: { outputModule: true },
};

/**
 * @param {string[]} targets
 * @returns {WebpackConfig[]}
 */
const getConfigs = (...targets) => {
    return targets.map((t) => ({
        ...Config_Base,
        name: t,
        target: t,
        output: {
            filename: "[name].js",
            path: path.resolve(__dirname, "dist", t),
            library: { type: "module" },

            // Explicitly enable for "webworker".
            chunkFormat: "module",
        },

        // Always `require("vscode")`.
        externals: { vscode: /^node(\d+(\.\d+)?)?$/.test(t) ? "node-commonjs vscode" : "commonjs vscode" },
    }));
};

module.exports = getConfigs("webworker", "node16");
```

</details>
<details>
<summary>Bundle: dist/node16/main.js</summary>

```javascript
import { createRequire as __WEBPACK_EXTERNAL_createRequire } from "module";
/******/ var __webpack_modules__ = ([
/* 0 */,
/* 1 */
/***/ ((module) => {

module.exports = __WEBPACK_EXTERNAL_createRequire(import.meta.url)("vscode");

/***/ }),
/* 2 */
/***/ ((__unused_webpack_module, __webpack_exports__, __webpack_require__) => {

__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   "hello": () => (/* binding */ hello),
/* harmony export */   "nice": () => (/* binding */ nice)
/* harmony export */ });
const hello = "Hello.";
const nice = "Nice!";


/***/ })
/******/ ]);
/************************************************************************/
/******/ // The module cache
/******/ var __webpack_module_cache__ = {};
/******/
/******/ // The require function
/******/ function __webpack_require__(moduleId) {
/******/ 	// Check if module is in cache
/******/ 	var cachedModule = __webpack_module_cache__[moduleId];
/******/ 	if (cachedModule !== undefined) {
/******/ 		return cachedModule.exports;
/******/ 	}
/******/ 	// Create a new module (and put it into the cache)
/******/ 	var module = __webpack_module_cache__[moduleId] = {
/******/ 		// no module.id needed
/******/ 		// no module.loaded needed
/******/ 		exports: {}
/******/ 	};
/******/
/******/ 	// Execute the module function
/******/ 	__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/
/******/ 	// Return the exports of the module
/******/ 	return module.exports;
/******/ }
/******/
/************************************************************************/
/******/ /* webpack/runtime/compat get default export */
/******/ (() => {
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = (module) => {
/******/ 		var getter = module && module.__esModule ?
/******/ 			() => (module['default']) :
/******/ 			() => (module);
/******/ 		__webpack_require__.d(getter, { a: getter });
/******/ 		return getter;
/******/ 	};
/******/ })();
/******/
/******/ /* webpack/runtime/define property getters */
/******/ (() => {
/******/ 	// define getter functions for harmony exports
/******/ 	__webpack_require__.d = (exports, definition) => {
/******/ 		for(var key in definition) {
/******/ 			if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 				Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 			}
/******/ 		}
/******/ 	};
/******/ })();
/******/
/******/ /* webpack/runtime/hasOwnProperty shorthand */
/******/ (() => {
/******/ 	__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
/******/ })();
/******/
/******/ /* webpack/runtime/make namespace object */
/******/ (() => {
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = (exports) => {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/ })();
/******/
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   "activate": () => (/* binding */ activate)
/* harmony export */ });
/* harmony import */ var vscode__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(1);
/* harmony import */ var vscode__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(vscode__WEBPACK_IMPORTED_MODULE_0__);
/* harmony import */ var _sub__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(2);


function activate(context) {
    vscode__WEBPACK_IMPORTED_MODULE_0__.window.showInformationMessage(_sub__WEBPACK_IMPORTED_MODULE_1__.hello);
}

})();

var __webpack_exports__activate = __webpack_exports__.activate;
export { __webpack_exports__activate as activate };
```

</details>
<details>
<summary>Bundle: dist/webworker/main.js</summary>

```javascript
/******/ var __webpack_modules__ = ([
/* 0 */,
/* 1 */
/***/ ((module) => {

module.exports = require("vscode");

/***/ }),
/* 2 */
/***/ ((__unused_webpack_module, __webpack_exports__, __webpack_require__) => {

__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   "hello": () => (/* binding */ hello),
/* harmony export */   "nice": () => (/* binding */ nice)
/* harmony export */ });
const hello = "Hello.";
const nice = "Nice!";


/***/ })
/******/ ]);
/************************************************************************/
/******/ // The module cache
/******/ var __webpack_module_cache__ = {};
/******/
/******/ // The require function
/******/ function __webpack_require__(moduleId) {
/******/ 	// Check if module is in cache
/******/ 	var cachedModule = __webpack_module_cache__[moduleId];
/******/ 	if (cachedModule !== undefined) {
/******/ 		return cachedModule.exports;
/******/ 	}
/******/ 	// Create a new module (and put it into the cache)
/******/ 	var module = __webpack_module_cache__[moduleId] = {
/******/ 		// no module.id needed
/******/ 		// no module.loaded needed
/******/ 		exports: {}
/******/ 	};
/******/
/******/ 	// Execute the module function
/******/ 	__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/
/******/ 	// Return the exports of the module
/******/ 	return module.exports;
/******/ }
/******/
/************************************************************************/
/******/ /* webpack/runtime/compat get default export */
/******/ (() => {
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = (module) => {
/******/ 		var getter = module && module.__esModule ?
/******/ 			() => (module['default']) :
/******/ 			() => (module);
/******/ 		__webpack_require__.d(getter, { a: getter });
/******/ 		return getter;
/******/ 	};
/******/ })();
/******/
/******/ /* webpack/runtime/define property getters */
/******/ (() => {
/******/ 	// define getter functions for harmony exports
/******/ 	__webpack_require__.d = (exports, definition) => {
/******/ 		for(var key in definition) {
/******/ 			if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 				Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 			}
/******/ 		}
/******/ 	};
/******/ })();
/******/
/******/ /* webpack/runtime/hasOwnProperty shorthand */
/******/ (() => {
/******/ 	__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
/******/ })();
/******/
/******/ /* webpack/runtime/make namespace object */
/******/ (() => {
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = (exports) => {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/ })();
/******/
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   "activate": () => (/* binding */ activate)
/* harmony export */ });
/* harmony import */ var vscode__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(1);
/* harmony import */ var vscode__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(vscode__WEBPACK_IMPORTED_MODULE_0__);
/* harmony import */ var _sub__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(2);


function activate(context) {
    vscode__WEBPACK_IMPORTED_MODULE_0__.window.showInformationMessage(_sub__WEBPACK_IMPORTED_MODULE_1__.hello);
}

})();

var __webpack_exports__activate = __webpack_exports__.activate;
export { __webpack_exports__activate as activate };
```

</details>

**Comment by savetheclocktower** [comment]
Having stumbled across [electron#21457](https://github.com/electron/electron/issues/21457) while investigating why an Atom extension couldn't use ESM imports, I'm a bit flummoxed. The crux of it seems to be that (a) Electron is complicated, (b) supporting ESM within Electron is complicated due to security issues and clashes in resolution algorithms between Node and browser environments; and (c) nobody seems to be sure what to do about it, including Electron's maintainers. The result is an issue that's been open for two years, is too important to be closed as a won't-fix, but seems no closer to resolution than on the day it was opened.

In a vacuum I'd take it as a symptom of a loss of mindshare and relevance — people seem much less sanguine about Electron than they used to be, and Atom (Electron's original reason for being) feels like it's basically in maintenance mode now. But then I remembered about Slack, and Figma, and Discord, and VSCode, and I'm now wondering how the platform underpinning all of these major apps is so paralyzed over ES modules. Is [electron#21457](https://github.com/electron/electron/issues/21457) a blocker for this issue, as @Silic0nS0ldier suspects?

The Node ecosystem will only embrace ESM more and more. For too long, publishing packages on NPM has involved module bundlers and transpilation and some amount of voodoo, but many package authors did it anyway, because there was the promise of an eventual blue-sky future when all of that would be unnecessary. Without ESM support in Electron, I imagine the only recourse for VSCode extension authors (like Atom package authors) is to resort to those same transpilers, and/or tools like [standard-things/esm](https://github.com/standard-things/esm) which were always intended as transitional stopgaps and will only work less and less well over time.

So I’m curious about whether [electron#21457](https://github.com/electron/electron/issues/21457) is on VSCode's radar, and whether its developers (and Slack's, and Discord's…) can perhaps be part of the brainstorming needed to get it moving.

The mystery I'm stuck on is why this impasse doesn't feel more urgent to major Electron apps like VSCode, but the answer could be quite simple. Perhaps there are aspects of VSCode's internals that make this problem less salient; or even remove Electron as a blocker; I'm not an expert there. But when VSCode is asked “Do you support ES modules?” in the year 2022, even a “yes, but…“ answer is a red flag when the answer is a flat “yes” most everywhere else.

**Comment by lewisl9029** [comment]
I started a new building a new extension a while back, here's the workaround I'm using at the moment that doesn't involve any third party libraries/transpilers:

In package.json, we point to a .cjs file as the main entrypoint:

```
  "main": "./main.cjs",
  "type": "module",
```

In main.cjs, we use dynamic import to import the ESM module, and then await it in activate:

```js
const importingMain = import('./main.js')

module.exports = {
  activate: async (context) => {
    ;(await importingMain).activate(context)
  },
}
```

In main.js, we use createRequire to import vscode which is still CommonJS, but otherwise everything else can be in native esm downstream:

```js
import { createRequire } from 'module'
const require = createRequire(import.meta.url)

const vscode = require('vscode')

export const activate = (context) => {
  //...
}
```

The setup seems to be working OK for me so far, hope it's helpful for others as well!

**Comment by Silic0nS0ldier** [comment]
@lewisl9029 I tried using your dynamic approach (similar to what I tried prior to posting back in December https://github.com/microsoft/vscode/issues/130367#issuecomment-987414840) without success.

First issue, `vsce` interprets `./main.cjs` as `./main.cjs.js`. Easily enough worked around by shifting ESM code into its own folder which contains `package.json` with `"type": "module"`, or using `.mjs`.

Second issue, I get an error once the dynamic import is hit. Same I hit previously.
`TypeError [ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING]: A dynamic import callback was not specified.`
Most search hits for this will turn up Jest which doesn't support ESM in a lot of cases (it replaces the module loader in CJS code).

Has anyone else had success? And if so, could you share your VSCode version and OS? Are you using VSCE to package your extension?

**Comment by lewisl9029** [comment]
@Silic0nS0ldier I haven't ran in to the first issue, but glad you found a workaround.

Re: the second issue, I seem to remember testing a packaged version with this pattern and seeing it work, but my memory must be failing me, because when I try to repro now I also see the same `ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING` error.

I traced it back to [the options to vm.Script](https://github.com/microsoft/vscode-loader/blob/main/src/loader.js#L831) missing a `importModuleDynamically` function to implement dynamic import resolution. 

I suppose if we have someone sufficiently motivated, and if we can get VSCode team's blessing, we could implement that function to get this pattern to work with packaged extensions. But I suspect there would be a high bar there in terms of security/compatibility implications and won't become available for quite a few releases.

In the mean time though, unfortunately it does mean this is pattern is only useful for local dev, which practically means not very useful at all. Apologies for getting people's hopes up only to disappoint. 😞

**Comment by Silic0nS0ldier** [comment]
@lewisl9029 Could be that support was removed in a past release, likely involuntarily. Regardless knowing that dynamic import works in the extension development host is useful information, it means Electron isn't the blocker.

**Comment by lewisl9029** [comment]
Found another semi-blocker to using native unbundled ESM: https://code.visualstudio.com/api/working-with-extensions/bundling-extension

> The first reason to bundle your Visual Studio Code extension is to make sure it works for everyone using VS Code on any platform. Only bundled extensions can be used in VS Code for Web environments like [github.dev](https://github.dev/) and [vscode.dev](https://vscode.dev/). When VS Code is running in the browser, it can only load one file for your extension so the extension code needs to be bundled into one single web-friendly JavaScript file. This also applies to [Notebook Output Renderers](https://code.visualstudio.com/api/extension-guides/notebook#notebook-renderer), where VS Code will also only load one file for your renderer extension.

This is a rather unfortunate constraint, because there are plenty of ways to flatten module waterfalls without bundling these days.

Unfortunately since I'm targeting Web as well, looks like I'm going to have to bundle regardless of ESM support. 😞

**Comment by Silic0nS0ldier** [comment]
I believe I've worked out why dynamic import is failing. Within the extension host a custom loader is used, [VSCode Loader](https://github.com/microsoft/vscode-loader), and [AMD](https://github.com/amdjs/amdjs-api/wiki/AMD) implementation.

The `NodeScriptLoader` (loader is built to work in multiple runtimes) loads modules using `require('vm').Script` which if not given a callback for option `importModuleDynamically` will cause loaded modules to error with `ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING`. [NodeJS docs](https://nodejs.org/api/vm.html#new-vmscriptcode-options)

To support dynamic import 2 things need to happen.

1. Provide a dynamic import callback (`importModuleDynamically`) in `NodeScriptLoader` (VSCode Loader).
   - https://github.com/microsoft/vscode-loader/blob/0f368c94f43a1616cc2867f9d2d5f724652b6f3b/src/core/scriptLoader.ts#L395
   - https://github.com/microsoft/vscode-loader/blob/0f368c94f43a1616cc2867f9d2d5f724652b6f3b/src/core/scriptLoader.ts#L470
2. Update the compiled loader script at `src/vs/loader.js` (VSCode).

Problems I can see here:
- Does NodeJS 14 support `importModuleDynamically`? It does not appear in `@types/node@14.14.43`.
- ES modules have different module resolution logic. I don't think `import()` can be used on its own (it resolve relative to the module scope). `import.meta.resolve` would help but is marked experimental.

**Comment by frank-dspeed** [comment]
Just my 5cent once you understood the both module systems your clear already able to use ESM only excempt as already saying is bootstraping of some electron code

as a Engine Coder it is clear why there is no ESM Support on bootstrap it is because NodeJS did not finish the Nativ Module bindings for the ESM Integration ok there is dlopen but thats it

## Short version
dynamic import() with full file path to a real ESM file will always work! and did work the last 7 years since it exists and did work 8+ years since it is in dev thats how long i use ESM inside Electron and vscode as also extensions just my 5cent.

there is a ESM Module Loader coded in CJS its called esm and is aviable via npm i esm 

It implements FULL! i Repeat FULL! More then Complet implementation of the ESM loader in Userland written in CJS. that even Works to bootstrap.

**Comment by Manish3323** [comment]
@frank-dspeed , you are correct. This issue seems to be more related to how tyescript works with dynamic imports. Stable version of `Typescript converts dynamic import() to require statements while doing transpilation`.
Typescript's beta seems to be trying to solve this issue. (https://github.com/microsoft/TypeScript/issues/46452)
`target: 'NodeNext'` is something i am looking forward to. This beta version still has some bugs. I am waiting for it to get stable.

**Comment by frank-dspeed** [comment]
@Manish3323 i am evaluating that a lot since some monthes and can tell you it solves nothing my final solution is simple but effecient i coded my own universal loader that can load ESM and CJS and invinted my own Module Standards that are universal compatible. 

The Main Problem between the both ecosystems is that ESM is Async by default while some environments do depend on something that is called a sideEffect and a sideEffect in the Nativ Environment means that you can not use ESM to catch the first events as it will get executed after the first process.nextTick also you can not require anything syncron from ESM as ESM is already Async by default even if the engine starts it it is Async. 

also all NodeNativ code can also only get required so all nativ code dependent node modules. can only get loaded via the CJS context that is bound in sync to the engine as it is Sync by default. 

if you want my conclusion about getting good support then the conlusion is the support is fully there while the coders are not aware of the edgecases like avoid default  exports and module.exports 
so go namedExports only if your forced to supply a default then simply always export a namespace with the same modules assigned to it that you export and also export the names this way the engine does no magic and you module works also transpiled with a single ts file. 

i publish the code of the loader soon it is in general rollup hooking into require to require also mjs files as cjs or mjs or vice versa on runtime and register the modules in the require cache

**Comment by DavidAnson** [comment]
I’ve hit this issue with the latest (0.48.0) release of the markdownlint extension: <https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint>. As one of the comments above notes, `import` works perfectly well during development (both for normal and web worker modes), but then fails after publishing a package to the Marketplace and using it. Unlike the people above that may be able to recompile or alter their code, my scenario involves loading a user script from their workspace at runtime, so I need `import` to function correctly. The error I get from VS Code is “Invalid host defined options”. I’m happy to provide more information or steps to reproduce if that would be helpful.

**Comment by fwouts** [comment]
@DavidAnson I was in a similar situation (my extension relies on https://github.com/vitejs/vite which recently migrated to ESM).

Here is the workaround I implemented:
- VS Code extension spins up a local Node server which contains all core logic (using https://github.com/sindresorhus/execa)
- VS Code extension communicates with the Node server via HTTP (ideally it would be via IPC, but I couldn't quite figure that out especially since `node-ipc` was recently compromised)

It's slower and uses more memory, but at least it doesn't crash 😕

See the code here: https://github.com/fwouts/previewjs/blob/a1222788f95b58cbbade800a4adbf584990fc291/integrations/vscode/src/index.ts#L40

**Comment by DavidAnson** [comment]
@fwouts, thank you so much! In my case, there has been very little adoption of .mjs configuration files (it's a new feature of the CLI), so I removed the documentation for this in the extension and will leave it to fail at runtime for anyone who tries. My hope is that VS Code will address the issue soon and it will start working without any further action.

**Comment by lucono** [comment]
I ran into this issue with the latest release of [Karma Test Explorer](https://marketplace.visualstudio.com/items?itemName=lucono.karma-test-explorer) while trying to update some dependencies that had now migrated to ESM. I was able to make it work as an ESM package by using esbuild to bundle to `cjs`, and bundling the output files with a `.cjs` extension as seen in the esbuild config [here](https://github.com/lucono/karma-test-explorer/blob/2106bb34d608e8bf38ba09a9d06a43d00e6b847e/esbuild.config.js#L5).

Also had [to alias](https://github.com/lucono/karma-test-explorer/blob/2106bb34d608e8bf38ba09a9d06a43d00e6b847e/esbuild.config.js#L15) the entry `main.cjs` file to `main.cjs.js`, which got it past a `vsce` packaging hiccup where it would look for `main.cjs.js` and not recognize the `main.cjs` in the `package.json` as a cjs file.

**Comment by ShenHongFei** [comment]
@lucono I have submitted a pr to vsce for .cjs extension
https://github.com/microsoft/vscode-vsce/pull/740

**Comment by ShenHongFei** [comment]
Create file named `patches/@vscode__vsce@2.17.0.patch`
```diff
# support cjs entry `main: "./xxx.cjs"` in package.json
# https://github.com/microsoft/vscode-vsce/pull/740

diff --git a/out/package.js b/out/package.js
index 3711bf76db20fa2b351089a8c32ae8c437a9b7d9..f9da497b9925f24e4596a85cb41aa2cd2d43946e 100644
--- a/out/package.js
+++ b/out/package.js
@@ -693,7 +693,7 @@ class LaunchEntryPointProcessor extends BaseProcessor {
         }
     }
     appendJSExt(filePath) {
-        if (filePath.endsWith('.js')) {
+        if (filePath.endsWith('.js') || filePath.endsWith('.cjs')) {
             return filePath;
         }
         return filePath + '.js';
```

add following section in package.json
```json
    "pnpm": {
        "patchedDependencies": {
            "@vscode/vsce@2.17.0": "patches/@vscode__vsce@2.17.0.patch"
        }
    }
```

pnpm install patched `@vscode/vsce` package
```shell
pnpm install
```

**Comment by ilg-ul** [comment]
There were some new discussion in the electron project, and I tried to make them aware that this problem is also important for the VS Code extensions: https://github.com/electron/electron/issues/21457#issuecomment-1458926542

There was also a more technical comment (https://github.com/electron/electron/issues/21457#issuecomment-1460028943), which I'm not able to reply, so, it would be great if someone with a better understanding of the problem could also join the discussion, perhaps we can move things forward.

**Comment by igorskyflyer** [comment]
Reading the latest [mention](https://github.com/electron/governance/issues/536#issuecomment-1465061802) of this issue - by the Electron team - the problem is _NOT_ in the Electron itself but in VS Code? Any thoughts on this? 🤔

**Comment by erickzhao** [comment]
> the latest https://github.com/electron/governance/issues/536#issuecomment-1465061802 of this issue - by the Electron team - the problem is NOT in the Electron itself but in VS Code

Hey @igorskyflyer, I wanted to clarify that I was merely referencing a previous comment from @Silic0nS0ldier in this GitHub issue: https://github.com/microsoft/vscode/issues/130367#issuecomment-1030721392.

I'm definitely not providing any new analysis on the topic (especially not in an "Electron team confirms it" way). I don't know how VSCode extensions are implemented and can't pretend that I know enough to know where the error should be fixed.

On another note, https://github.com/electron/electron/pull/37535 (currently under review) does expand ESM support in Electron, which might be promising to the VSCode team eventually. :)

**Comment by igorskyflyer** [comment]
@erickzhao, thank you for your insightful comment and for clearing things out! I hope this issue will be resolved soon, we would all benefit from it. 😁

**Comment by JustinGrote** [comment]
This was just added to Electron, the question will be when Electron 28 gets merged into vscode. If a tracker issue for that ever gets created then this issue should be closed/tied to that dependent issue.
![image](https://github.com/microsoft/vscode/assets/15258962/e53c07be-772d-49d4-8dce-f702a1bd51cd)

**Comment by felipecrs** [comment]
I would say there is a long way to go... vscode stable still uses Electron 22, and latest stable Electron is still 26.

**Comment by JustinGrote** [comment]
@felipecrs right, I'm not saying it's happening tomorrow, but the downstream blocker is now removed so now it's entirely on the vscode team to implement rather than it being kicked to a dependency :)

**Comment by felipecrs** [comment]
For sure. I just wanted to set the expectations. Thanks for the heads up.

**Comment by felipecrs** [comment]
I have a honest question, is there any reason why VS Code doesn't keep up more closely with Electron updates? Is it only because it's too hard to handle the breaking changes?

**Comment by liuxingbaoyu** [comment]
FYI, for some cases it is possible to use `worker_threads` to use the esm package, prettier-vscode is doing this.
https://github.com/prettier/prettier-vscode/pull/3016#issue-1744454011

**Comment by loynoir** [comment]
@liuxingbaoyu 

Wow! Awesome workaround!

**Comment by tats-u** [comment]
https://github.com/electron/electron/issues/21457

https://www.electronjs.org/docs/latest/tutorial/electron-timelines

Electron 28 will be available in November.
VS Code uses the versions $3n + 1$ of Electron after 19. (19→22→25) The next will be 28.

https://github.com/microsoft/vscode/pull/154092
https://github.com/microsoft/vscode/pull/177600
https://github.com/microsoft/vscode/pull/188268

The lag is about 2 months.

I hope this will be fixed at the beginning of the next year.

**Comment by igorskyflyer** [comment]
> [electron/electron#21457](https://github.com/electron/electron/issues/21457)
> 
> https://www.electronjs.org/docs/latest/tutorial/electron-timelines
> 
> Electron 28 with the fix will be available on November. VS Code uses the versions 3n+1 of Electron after 19. (19→22→25) The next will be 28.
> 
> #154092 #177600 #188268
> 
> The lag is about 2 months.
> 
> I hope this will be fixed in VS Code at the beginning of the next year.

Wow, this is actually great news, I hope so too! We can then finally switch to ESM, use the standardized ESM syntax and extensions will then get smaller = faster because of tree-shaking. Thanks for the info. 😁

**Comment by justinfagnani** [comment]
VS Code just updating Electron probably isn't automatically going to allow for standard JS module extension, and it probably isn't necessary since current VS Code can already load JS modules via dynamic `import()`.

VS Code `require()`s extensions, so that will have to be changed to a dynamic `import()`, which could already be done today.

And for our own extension code, it's relatively easy to make the extension package `type: "module"` in package.json and have one `extension.cjs` file that dynamic imports the rest of the extension in `activate()` since activate returns a Promise.

**Comment by JustinGrote** [comment]
`And for our own extension code, it's relatively easy to make the extension package type: "module" in package.json and have one extension.cjs file that dynamic imports the rest of the extension in activate() since activate returns a Promise.`

That would be after they make it an `import` though right? Do you have an example of an extension doing this today?

**Comment by justinfagnani** [comment]
> That would be after they make it an `import` though right

No, any script can use dynamic import. The tricky part is that dynamic imports are async, so you need to call them from other async entrypoints, but activate is already that.

> Do you have an example of an extension doing this today?

I have it working locally. It looks like:

`extension.cts`:
```ts
import * as vscode from 'vscode';

export async function activate(context: vscode.ExtensionContext) {
  const {FooEditorProvider} = await import('./lib/foo-editor.js');
  context.subscriptions.push(FooEditorProvider.register(context));
}
```

`lib/foo-editor.ts`:
```ts
import * as vscode from 'vscode';

export class FooEditorProvider implements vscode.CustomTextEditorProvider {
  public static register(context: vscode.ExtensionContext): vscode.Disposable {
    // ...
  }
}
```

**Comment by kungfooman** [comment]
I described my pure ESM solution here: https://github.com/xenova/transformers.js/issues/317#issuecomment-1732543840

Pretty much like yours @justinfagnani while using `package.json` as sorts of importmap to circumvent that you cannot do a normal `import * as vscode from 'vscode'` in ESM.

**Comment by AlfishSoftware** [comment]
> VS Code just updating Electron probably isn't automatically going to allow for standard JS module extension, and it probably isn't necessary since current VS Code can already load JS modules via dynamic import().

What about Web extensions (extensions that can run in browser IDEs like vscode.dev and github.dev)?
Would this update allow us to have multiple module files instead of having to use webpack? From [the docs](https://code.visualstudio.com/api/extension-guides/web-extensions#web-extension-main-file):
> [...] The script runs in the web extension host in a [Browser WebWorker](https://developer.mozilla.org/docs/Web/API/Web_Workers_API) environment. It is restricted by the browser worker sandbox and has limitations compared to normal extensions running in a Node.js runtime.
> - Importing or requiring other modules is not supported. `importScripts` is not available as well. As a consequence, the code must be packaged to a single file.
> - The [VS Code API](https://code.visualstudio.com/api/references/vscode-api) can be loaded via the pattern require('vscode'). This will work because there is a shim for require, but this shim cannot be used to load additional extension files or additional node modules. It only works with require('vscode').

**Comment by tats-u** [comment]
https://github.com/microsoft/vscode/issues/196941
The next Electron is 27 not 28...

**Comment by JustinGrote** [comment]
As an update here, until this is supported natively, I've had pretty good success with both esbuild and packer/swc in converting ESM to CJS automatically during bundling, including synthetic default imports and whatnot.

**Comment by theoludwig** [comment]
> #196941 The next Electron is 27 not 28...

The next one is now Electron v28.
Progress can be tracked here: https://github.com/microsoft/vscode/issues/201935

We should be able to close this issue very soon! :smile:

**Comment by JustinGrote** [comment]
> I personally am desperate to make typescript + vite work for VSCode extension development.
> 
> It seems that any `tsconfig.json` (https://github.com/Malix-off/VSC-Extension-StashFlow/blob/dev/tsconfig.json) and `vite.config` (https://github.com/Malix-off/VSC-Extension-StashFlow/blob/dev/vite.config.ts) I tried didn't worked
> 
> I think it might be tied to that issue. Do you guys found a workaround?
> 
> ## Update:
> 1. ESNext TS + Vite works now!
>    Thanks a lot to @rxliuli for [chore: Update vite config of build cjs output Malix-off/VSC-Extension-StashFlow#4](https://github.com/Malix-off/VSC-Extension-StashFlow/pull/4) ❤️

FYI pretty sure this won't work inside a vscode runner with vscode-test, because the vitest runner is ES only.

**Comment by Malix-off** [comment]
I didn't use vitest, so can't confirm

**Comment by starball5** [comment]
Related on Stack Overflow: [How do I import an ES6 JavaScript module in my VS Code extension written in TypeScript?](https://stackoverflow.com/q/70620025/11107541)

**Comment by alexandruavadanii** [comment]
Looks like https://github.com/microsoft/vscode/pull/203956 was just merged in `main`, so the next Insiders build will be based on Electron 28.

**Comment by igorskyflyer** [comment]
> Looks like #203956 was just merged in `main`, so ~the next~ one of the next Insiders builds will be based on Electron 28 (probably after the endgame release from the stable branch is verified).

<div align="center">
 
<img src="https://github.com/microsoft/vscode/assets/20957750/b98b11cc-eac4-4c1b-aca5-07b8bc3ac815" alt="Igor Dimitrijević (igorskyflyer) shared an image.">
</div>

**Comment by alexandruavadanii** [comment]
I'm not familiar enough with js, so I don't know how much of this is still relevant, but an issue has been waiting in the vscode-loader repo at https://github.com/microsoft/vscode-loader/issues/36 for a while. Maybe others do have expertise in that area and can chime in.

**Comment by lppedd** [comment]
So the answer is we're still mid-way. The base platform now supports ES modules, but the VS Code infrastructure doesn't.  
Correct?

**Comment by lppedd** [comment]
Btw, there is also the question of Web Extensions to consider.

IIRC as of now we compile them using CJS. And I suppose the base framework needs to be updated to handle ESM imports.

**Comment by metawrap-dev** [comment]
I have a project that depends on being able to load ESM into VSCode extensions, which I assume also means that VSCode extensions themselves need to be ESM. 

I'm considering downgrading the project to CJS. 

Should I wait? Is a fix to this weeks away or months... or more?

**Comment by itsdouges** [comment]
Hi folks — can you confirm dynamic imports _should_ work in VSCode extensions? When running my code for production I'm getting an `Invalid host defined options` error:

```
[error] TypeError: Invalid host defined options
    at Object.module.exports.<computed> [as createServer] (/Users/douges/.vscode/extensions/triplex.triplex-vscode-0.0.1/node_modules/vite/index.cjs:23:36)
```

Which, when zooming to that line of code, is this highlighted line:

```diff
// index.cjs
asyncFunctions.forEach((name) => {
  module.exports[name] = (...args) =>
+    import('./dist/node/index.js').then((i) => i[name](...args))
})
```

Leading me to believe it's not supported. Could it be because Vite has this in a `cjs` file? Something else?

**Comment by lppedd** [comment]
@metawrap-dev imo you should switch to CJS.

Don't make plans based on what you read on issues. Not the first time highly requested features take years to land.

**Comment by JustinGrote** [comment]
@itsdouges @metawrap-dev  as of today, the general recommendation is to use a bundler like `esbuild` (recommended) or `rollup` and use their features/plugins to convert ES modules to CJS automatically.

Generally have had no issues with this, the only major roadblock is you can't run E2E in-vscode testing using testing frameworks that require ESM such as `vitest`, but `mocha` is the de-facto standard used by everything including the official vscode test cli and you can still run tests that don't need/reference the vscode api or mock it in whatever framework you want.

**Comment by itsdouges** [comment]
I understand. Unfortunately there are a few packages that don’t work well
with bundling, so I’m doing both: bundling but marking dependencies that
can’t be bundled as external.

**Comment by metawrap-dev** [comment]
> @itsdouges @metawrap-dev as of today, the general recommendation is to use a bundler like `esbuild` (recommended) ...

Thanks. That's exactly what I did as a workaround. I am glad that you have confirmed the choice,

**Comment by v4lentin1879** [comment]
Can anyone tell me, is there any possibility of importing a esm dependency in a vscode extension?

**Comment by lppedd** [comment]
You can import an ESM package, but you then have to bundle as CJS. This makes the whole experience considerably worse, as the restart cycles during development are longer, and most of the times source maps don't even work.

**Comment by tats-u** [comment]
Why is this in not "On Deck" but "Backlog"?

**Comment by tats-u** [comment]
There has ever been mentioning the concrete error message.

An extension using Pyodide:

https://github.com/microsoft/vscode/issues/197329

Prettier (v10.0.0):

```
TypeError: A dynamic import callback was not specified.
	at importModuleDynamicallyCallback (node:internal/modules/esm/utils:228:9)
	at Object.<anonymous> (c:\Users\tatsu\.vscode\extensions\esbenp.prettier-vscode-10.0.0\node_modules\prettier\index.cjs:600:23)
	at u._compile (c:\Users\tatsu\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:2:1257)
	at Module._extensions..js (node:internal/modules/cjs/loader:1432:10)
	at Module.load (node:internal/modules/cjs/loader:1215:32)
	at Module._load (node:internal/modules/cjs/loader:1031:12)
	at c._load (node:electron/js2c/node_init:2:13801)
	at E._load (c:\Users\tatsu\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:177:6051)
*snip*
```

```
 600   │ var prettierPromise = import("./index.mjs");
```

**Comment by tats-u** [comment]
In VS Code Exploration available from https://github.com/microsoft/vscode/issues/160416#issuecomment-2292791157, the error message is changed:
The execution of `.mjs` file itself looks successful.

```
Cannot find module 'c:\Users\WDAGUtilityAccount\.vscode-exploration\extensions\esbenp.prettier-vscode-10.0.0\node_modules\prettier\internal\internal.mjs' imported from c:\Users\WDAGUtilityAccount\.vscode-exploration\extensions\esbenp.prettier-vscode-10.0.0\node_modules\prettier\index.mjs
Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'c:\Users\WDAGUtilityAccount\.vscode-exploration\extensions\esbenp.prettier-vscode-10.0.0\node_modules\prettier\internal\internal.mjs' imported from c:\Users\WDAGUtilityAccount\.vscode-exploration\extensions\esbenp.prettier-vscode-10.0.0\node_modules\prettier\index.mjs
    at finalizeResolution (node:internal/modules/esm/resolve:265:11)
    at moduleResolve (node:internal/modules/esm/resolve:940:10)
    at defaultResolve (node:internal/modules/esm/resolve:1176:11)
    at nextResolve (node:internal/modules/esm/hooks:866:28)
    at resolve (data:text/javascript;base64,CglleHBvcnQgYXN5bmMgZnVuY3Rpb24gcmVzb2x2ZShzcGVjaWZpZXIsIGNvbnRleHQsIG5leHRSZXNvbHZlKSB7CgkJaWYgKHNwZWNpZmllciA9PT0gJ2ZzJykgewoJCQlyZXR1cm4gewoJCQkJZm9ybWF0OiAnYnVpbHRpbicsCgkJCQlzaG9ydENpcmN1aXQ6IHRydWUsCgkJCQl1cmw6ICdub2RlOm9yaWdpbmFsLWZzJwoJCQl9OwoJCX0KCgkJLy8gRGVmZXIgdG8gdGhlIG5leHQgaG9vayBpbiB0aGUgY2hhaW4sIHdoaWNoIHdvdWxkIGJlIHRoZQoJCS8vIE5vZGUuanMgZGVmYXVsdCByZXNvbHZlIGlmIHRoaXMgaXMgdGhlIGxhc3QgdXNlci1zcGVjaWZpZWQgbG9hZGVyLgoJCXJldHVybiBuZXh0UmVzb2x2ZShzcGVjaWZpZXIsIGNvbnRleHQpOwoJfQ==:13:10)
    at nextResolve (node:internal/modules/esm/hooks:866:28)
    at Hooks.resolve (node:internal/modules/esm/hooks:304:30)
    at handleMessage (node:internal/modules/esm/worker:196:24)
    at Immediate.checkForMessages (node:internal/modules/esm/worker:138:28)
    at process.processImmediate (node:internal/timers:478:21)
```

**Comment by tats-u** [comment]
If the update in the next month doesn't help, we need to wait for `vm.Module`. https://github.com/nodejs/node/issues/37648

**Comment by mark-wiemer** [comment]
For SEO, here is a similar error message I've received for an LSP extension:

```log
2024-09-28 11:15:57.990 [info] ExtensionService#_doActivateExtension mark-wiemer.vscode-autohotkey-plus-plus, startup: false, activationEvent: 'onLanguage:ahk2'
2024-09-28 11:15:58.013 [error] Activating extension mark-wiemer.vscode-autohotkey-plus-plus failed due to an error:
2024-09-28 11:15:58.013 [error] Error [ERR_REQUIRE_ESM]: require() of ES Module c:\Users\markw\.vscode\extensions\mark-wiemer.vscode-autohotkey-plus-plus-6.1.2\dist\extension.js from c:\Users\markw\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js not supported.
Instead change the require of extension.js in c:\Users\markw\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js to a dynamic import() which is available in all CommonJS modules.
    at c._load (node:electron/js2c/node_init:2:13801)
    at y._load (c:\Users\markw\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:184:6031)
    at i._load (c:\Users\markw\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:181:32144)
    at n._load (c:\Users\markw\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:146:34326)
    at Module.require (c:\Users\markw\.vscode\extensions\github.copilot-chat-0.20.3\dist\extension.js:48:34588)
    at i (c:\Users\markw\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:3:98)
    at m.xb (c:\Users\markw\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:181:33345)
    at async Promise.all (index 0)
```

**Comment by tats-u** [comment]
Hmm, we should wait `vm.Modules`.... https://github.com/nodejs/node/issues/37648
Possibly `require(esm)`  might also be an escape hatch:
https://github.com/nodejs/node/pull/55085
https://github.com/nodejs/node/pull/55217

**Comment by yxw007** [comment]
Latest vscode I use esm works fine 🎉🎉🎉

**Comment by typed-sigterm** [comment]
> Latest vscode I use esm work fine 🎉🎉🎉

Unfortunately `"type": "module"` hasn't been supported yet

**Comment by CoffeeChaton** [comment]
> Latest vscode I use esm work fine 🎉🎉🎉

Which repositories?

This target is cjs
https://github.com/yxw007/vscode-translate-next/blame/c61d60d16d86b0d674a35314aeaa54b05c6b1c22/esbuild.js#L32-L33

**Comment by JustinGrote** [comment]
> > Latest vscode I use esm work fine 🎉🎉🎉
> 
> Which repositories?
> 
> This target is cjs [yxw007/vscode-translate-next@`c61d60d`/esbuild.js#L32-L33 (blame)](https://github.com/yxw007/vscode-translate-next/blame/c61d60d16d86b0d674a35314aeaa54b05c6b1c22/esbuild.js#L32-L33)

Agreed, very important to clarify using ESM modules and compiling to CJS is *not* the same as using them "natively"

**Comment by igorskyflyer** [comment]
> > > Latest vscode I use esm work fine 🎉🎉🎉
> > 
> > 
> > Which repositories?
> > This target is cjs [yxw007/vscode-translate-next@`c61d60d`/esbuild.js#L32-L33 (blame)](https://github.com/yxw007/vscode-translate-next/blame/c61d60d16d86b0d674a35314aeaa54b05c6b1c22/esbuild.js#L32-L33)
> 
> Agreed, very important to clarify using ESM modules and compiling to CJS is _not_ the same as using them "natively"

@yxw007,
this was always possible... Currently, only VS Code itself runs with ESM, extensions still don't allow it.

**Comment by yxw007** [comment]
@igorskyflyer @CoffeeChaton @JustinGrote 
I just changed the packaging results to esm for testing, and found that it cannot be used normally, but it is still possible to use esm in the code.

**Comment by JustinGrote** [comment]
> @igorskyflyer @CoffeeChaton @JustinGrote 
> I just changed the packaging results to esm for testing, and found that it cannot be used normally, but it is still possible to use esm in the code.

You need to be very careful what you are saying. You are not "using esm", you are *transpiling to commonjs*, it's a big difference.

**Comment by NullVoxPopuli** [comment]
So can we use untranspiled ESM?
like, if I author extension in JS (in ESM), I should be able to ship it without compiling?

**Comment by JustinGrote** [comment]
@NullVoxPopuli not yet, while vscode itself is using ESM now, the extension engine is still bootstrapped via AMD, so extensions still have a commonjs entrypoint and thus need to continue to be commonjs when compiled.

**Comment by felipecrs** [comment]
Maybe someone can clarify what is the game plan to get ESM support for extensions?

**Comment by JustinGrote** [comment]
It has to be fixed in the vscode-loader repository is my current understanding, here is the upstream issue for that: https://github.com/microsoft/vscode-loader/issues/56

**Comment by ilg-ul** [comment]
>  the upstream issue for that: https://github.com/microsoft/vscode-loader/issues/56

That issue is dated February 29, 2024 and it doesn't seem very active. :-(

Is this the only remaining issue to be fixed?

**Comment by axefrog** [comment]
@alexdima @jrieken Hey any chance you could weigh in on this? I've been Googling for about half an hour and it's quite hard to find any kind of definitive advice on what those of us implementing a VSCode extensions should be doing given how much of the world has moved on to ESM modules by default.

**Comment by JustinGrote** [comment]
@axefrog https://github.com/microsoft/vscode/issues/130367#issuecomment-1819362935
Just use esbuild or bundler of your choice to compile to commonJS, you can use pretty much all the ESM libraries out there for now, it's just not optimal till this gets resolved.

**Comment by NullVoxPopuli** [comment]
Compilation is easy.
I want to debug source without transpilation and sourcemaps

**Comment by JustinGrote** [comment]
> Compilation is easy. I want to debug source without transpilation and sourcemaps

VSCode makes sourcemaps really easy to work with so that you pretty much can't tell the difference, you can also use tools like esbuild/register to work with the ts files "directly" (they get transpiled on the fly in node)

**Comment by axefrog** [comment]
@JustinGrote Thankyou! Works beautifully. For anyone else arriving here, I'm using Bun.build for this. It is shockingly fast.

CLI:
```
bun build --entrypoints ./src/extension.ts --outdir ./out --target node --format cjs --sourcemap=linked --external vscode --watch
```

Documentation: https://bun.sh/docs/bundler

**Comment by bmingles** [comment]
There are now instructions in vscode docs for configuring esbuild:
https://code.visualstudio.com/api/working-with-extensions/bundling-extension#using-esbuild

This can also be auto configured when scaffolding a new extension using yeoman

**Comment by tats-u** [comment]
require(esm), one of the candidates for the escape hatch, is unflagged in the latest Node 22 (22.12).
Waiting for being merged into Electron and VS Code.

**Comment by JustinGrote** [comment]
> There are now instructions in vscode docs for configuring esbuild: [code.visualstudio.com/api/working-with-extensions/bundling-extension#using-esbuild](https://code.visualstudio.com/api/working-with-extensions/bundling-extension#using-esbuild)
> 
> This can also be auto configured as one of the options when scaffolding a new extension using yeoman

Just again want to make a note, because this is a thread about `esm`, these instructions are for using esbuild to **transpile ES modules to commonjs**, they do not provide native esm support.

**Comment by felipecrs** [comment]
For me, transpiling to commonjs with esbuild did not help. I'm still not able to consume ESM-only dependencies like newest version of `execa`.

But I was able to write ESM-like code at least, but that isn't very useful.

**Comment by igorskyflyer** [comment]
> For me, transpiling to `commonjs` with `esbuild` did not help. I'm still not able to consume ESM-only dependencies like newest version of `execa`.
> 
> I was able to write ESM-like code at least, but that isn't very useful.

@felipecrs, do you have a working repository for it so I can do some tests tomorrow? 😄

**Comment by felipecrs** [comment]
@igorskyflyer, yeah: https://github.com/vscode-shellcheck/vscode-shellcheck. My last attempt: https://github.com/vscode-shellcheck/vscode-shellcheck/pull/1555.

**Comment by igorskyflyer** [comment]
> @igorskyflyer, yeah: https://github.com/vscode-shellcheck/vscode-shellcheck. My last attempt: https://github.com/vscode-shellcheck/vscode-shellcheck/pull/1555.

Thanks, I'll check it out tomorrow. 🙂

**Comment by igorskyflyer** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/20957750?s=64&amp;u=0d3d6f7e9bd2397877eae54c8c1085b43e90ed65&amp;v=4">@igorskyflyer](https://github.com/igorskyflyer?rgh-link-date=2025-01-02T01%3A06%3A44.000Z), yeah: [vscode-shellcheck/vscode-shellcheck](https://github.com/vscode-shellcheck/vscode-shellcheck?rgh-link-date=2025-01-02T01%3A06%3A44.000Z). My last attempt: [vscode-shellcheck/vscode-shellcheck#1555](https://github.com/vscode-shellcheck/vscode-shellcheck/pull/1555).

@felipecrs, sorry for the delay, I was busy with personal stuff. Anyway, after upgrading `execa` and `minimatch` I got a few errors in the `tool-check.ts` and the `linter.ts` files. After fixing the given import errors, I got it to compile to a proper CommonJS (and a JS map file) but I cannot seem to test it, I get the error:  

> Error: there is no registered task type 'npm'. Did you miss installing an extension that provides a corresponding task provider?
> Activating task providers npm
> Error: Invalid problemMatcher reference: $esbuild-watch  

Even though I have npm installed and the extension working and running. Did you get to that point or you weren't even able to compile it?

**Comment by yne** [comment]
Just to clarify, it has been possible to import .mjs from a .js file, here is a minimal example:

`lib.mjs`
```js
export default 42
```
`main.js` (.mjs files are refused even if they have no import/export statements)
```js
exports.activate = async function () {
    console.log((await import("./lib.mjs")).default)
}
```
`package.json`
```json
{
	"name": "basic",
	"version": "0.1.0",
	"main": "./main.js",
	"engines": { "vscode": "^1.96.0" },
	"activationEvents": [ "onStartupFinished" ]
}
```
Open the .js file, run > start debugging > Vscode Extension, and you shall see a `42` in your host vscode's debug console

However, As @JustinGrote said, the goal is to bootstram .mjs extension. and the upstream

So I tried on 1.96 the following changes to the package.json, but with no success (and no surprises):

```diff
+	"type": "module",
+	"module": "./basic.mjs",
-	"main": "./main.js",
```
=> this does not even try to run basic.mjs

```diff
+	"type": "module",
 	"main": "./main.mjs",
```
=> this fail with `require() of ES Module basic.mjs not supported`

Hope this clarify the current state of the affair, because a lot of comment seems to be bundling/transpiling issues.

**Comment by felipecrs** [comment]
@igorskyflyer let's continue that conversation in https://github.com/vscode-shellcheck/vscode-shellcheck/pull/1555#issuecomment-2572083957 to avoid spamming other one's inboxes.

**Comment by jogibear9988** [comment]
is esm support for extensions something wich is acteively worked on? Will vscode at some time switch to deliver ESM code instead of "require()"?

**Comment by JustinGrote** [comment]
@jogibear9988 it's not on any official roadmap I'm aware of, but the core of VSCode was recently switched to use ESM native which was a tremedous effort. The current roadblock I see is that the vscode-loader has to be updated, that's tracked at this upstream issue:
https://github.com/microsoft/vscode-loader/issues/36

**Comment by itsdouges** [comment]
With Node.js supporting ESM through CJS hopefully the uplift to just enable extensions using ESM won't be that large? 🙏

**Comment by JanMiksovsky** [comment]
I wrote up a technique for [Writing a VS Code extension in ES modules in early 2025](https://jan.miksovsky.com/posts/2025/03-17-vs-code-extension.html) that confines the use of CommonJS to a single module that loads the ES. This works today without compilation/bundling and should be easy to migrate to using only ES when VS Code supports that.

**Comment by jogibear9988** [comment]
My extension also uses ESM modules already, but only in a CustomEdtorView (https://marketplace.visualstudio.com/items?itemName=node-projects.vscode-designer-addon)

**Comment by JustinGrote** [comment]
> I wrote up a technique for [Writing a VS Code extension in ES modules in early 2025](https://jan.miksovsky.com/posts/2025/03-17-vs-code-extension.html) that confines the use of CommonJS to a single module that loads the ES. This works today without compilation/bundling and should be easy to migrate to using only ES when VS Code supports that.

Yes dynamic imports are an OK workaround, but how would you handle the deactivate method in this case?

**Comment by eamodio** [comment]
@JanMiksovsky 

> You could pass it as a function parameter, but to keep things simpler, I just had extension.cjs set a global variable on globalThis so extension.mjs can read that global. I believe each extension runs in its own process; this should be safe enough

VS Code loads extensions into a single extension host process, so they are NOT isolated from one another.

**Comment by jrieken** [comment]
With https://github.com/microsoft/vscode/pull/245214 this has landed.

TL;DR

* this ONLY for the NodeJS extension host not yet supported for the web worker extension host
* this will land with next insiders, which is likely happing on April 3rd or 4th. 
* extensions need to use `type: module` or the `.mjs` suffix
* extension can use native `import`-statements _including_ `import('vscode')` and things should just work

I haven't had the chance to read through all comments but I things there two topics: how to transpile ESM code to CJS and how to natively support ESM for vscode extensions. The former was always possible and just a matter of configuring you favorite build tools (esbuild, tsc, etc...) What I pushed is support for the latter, e.g vscode uses `import` to load extensions and extensions can use `import` to load their dependencies, including the `vscode`-module

##### Why was this complicated? Why did this take so long?

The `vscode`-module is not a real module, e.g there is no such file on disk. When an extension calls `import('vscode')` (or `require('vscode')`) we create an extension-specific instance of the API. This ensures extensions can have different API proposals enabled and also ensures APIs know from which extension they are used, for instance this enables us to tell users what extension has opened a notification. Fortunately there is NodeJS API that allows us to support "module_is_not_real"-vehicle for ESM just like we did for CJS. 

##### What about the web?

The web worker extension host is a worker thread launched from an iframe. ESM in the web does support "import maps" which would allow to build a similar "module_is_not_real"-vehicle but AFAIK this is not supported in web workers. If anyone knows more about it or has another idea, speak up.

**Comment by estruyf** [comment]
Great job @jrieken!!! 💯

**Comment by beingminimal** [comment]
@abhi-parate-03

**Comment by TomasHubelbauer** [comment]
> If anyone knows more about it or has another idea, speak up.

@jrieken could a service worker be used to make up the module on the fly via request interception? The same way a service worker can be used to make up a web resource on the fly when serving offline-mode home page etc. Check this out for how to make service workers make up request: https://github.com/TomasHubelbauer/bun-bundle-service-worker/blob/main/worker.ts. TL;DR: use `new Response(madeUpBody)`

As an aside, this is amazing news - a bit iffy to drop this on April's Fools, it would break my little heart if this wasn't real! :D

**Comment by jrieken** [comment]
@TomasHubelbauer I did think a little about service workers and my main concerns are they are meant to be opaque, e.g they are just an optimisation and things should work without, they are bit finicky to manage esp. in different deployments like desktop, vscode.dev, gh.dev, and last I don't if the interception knows the origin of a request. For our tricks to work it is crucial to know what file/module makes an import-call. 


> I'm a bit skeptical about ESM adoption going forward given the missing web support tho.

Yeah, it's not because I don't want to solve this but because I don't how to do it. That said, web worker extensions are already constraint to be a single file (otherwise `require(vscode)` doesn't work) which means you already have a bundle/transpile pipeline in place

**Comment by TomasHubelbauer** [comment]
@jrieken Alternatively, vscode.dev could expose a `/vscode` endpoint serving a dynamic `text/javascript` response which would know about the user's identity, open project etc. and could generate the same response the service worker would have been crearting, but on the BE instead of the FE. Not sure if this is practical given the shift of the load to the service instead of the browser, but technically I believe it should be possible. Additionally, if there is a way of code-sharing this logic, the service worker would then become truly opaque - it could take care of serving `./vscode` if it is installed, but if it wasn't installed or supported, the actual server endpoint would take over. If your vscode.dev setup is such that code-sharing would be possible between the service worker implementation and the server implementation, maintenance could be reasonable - but this is hard to get right from my experience. AFAIK the service worker can access the request object information so when it comes to knowing the origin of the request to know what to serve exactly, this information could be also included in a cookie or a request header so that the service worker could use it to nail down exactly what it is serving.

**Comment by tats-u** [comment]
Electron 35 (Node.js 22.14, where require(ESM) is available): 

- https://github.com/microsoft/vscode/issues/245420

**Comment by justarandomgeek** [comment]
I've been experimenting with migrating my extension (https://github.com/justarandomgeek/vscode-factoriomod-debug/tree/next) to ESM, and so far it works *great* with one notable exception: a few of my deps are CJS still, and when they get rolled into my ESM bundle (by esbuild) they still try to `require("node:events")` (and various other node internal modules), which fails. For now i'm shimming this with some esbuild `external` imports of separate .cjs bundles of those specific deps to then load by the main bundle.

**Comment by beingminimal** [comment]
@abhi-parate-st3

**Comment by beingminimal** [comment]
cc @abhi-parate-st3

**Comment by justarandomgeek** [comment]
after doing a bit more research i've removed the cjs shims mentioned in my previous comment and instead just added 
```js
	banner: {
		// https://github.com/evanw/esbuild/issues/1232#issuecomment-830677608
		// shim `require` for cjs deps to load properly in esm-land
		js: `
		import {createRequire} from 'module'
		const require = createRequire(import.meta.url)
		`,
	},
```
in my esbuild bundling config, and now i've got pure-esm extension (and cli) bundle output, and everything seems to be working! Thanks for the great work everybody! :) 🎉

**Comment by lishaduck** [comment]
@justarandomgeek, you could try switching to rolldown (or tsdown), which handles that automatically.

**Comment by felipecrs** [comment]
I got this error when launching the extension:

```
Dynamic require of "child_process" is not supported.
```

I used [execa](https://github.com/sindresorhus/execa). Is this a limitation of ESM support?

**Comment by justarandomgeek** [comment]
@felipecrs that's the same issue i ran into with bundled CJS deps, see my last comment above for the shim you need to fix it - you need to pull in a `require` for that bundled CJS dep to load node internal modules!

**Comment by justarandomgeek** [comment]
Yeah most of mine came up via transitive deps, rather than direct deps. In any case, getting a `require` from `createRequire` should sort it right out!

I also experimented some with using esbuild `alias` entries to force things that load dual-mode packages via `require` to get the esm flavor instead of the cjs flavor from them (especially for a few where my bundles ended up with *both* flavors >.> ), but that made more mess than it helped!

**Comment by felipecrs** [comment]
@jrieken maybe you can amend your [sample repository](https://github.com/jrieken/vscode-esm-sample-extension) with bundling instructions?

**Comment by felipecrs** [comment]
@justarandomgeek thank you. The esbuild banner trick worked like a charm.

---

Now I'm stuck with the `import from "vscode"`. It works great when debugging the extension, but tests fail with:

```console
> vscode-test

✔ Validated version: 1.100.0
✔ Found at https://update.code.visualstudio.com/1.100.0/linux-x64/stable?released=true
✔ Downloaded VS Code into /home/felipecrs/repos/vscode-shellcheck/.vscode-test/vscode-linux-x64-1.100.0
[main 2025-05-10T20:13:20.294Z] update#setState disabled
[main 2025-05-10T20:13:20.296Z] update#ctor - updates are disabled by the environment
Started local extension host with pid 87078.
Started initializing default profile extensions in extensions installation folder. file:///home/felipecrs/repos/vscode-shellcheck/.vscode-test/extensions
ComputeTargetPlatform: linux-x64
Completed initializing default profile extensions in extensions installation folder. file:///home/felipecrs/repos/vscode-shellcheck/.vscode-test/extensions
Loading development extension at /home/felipecrs/repos/vscode-shellcheck
Error: Cannot find package 'vscode' imported from /home/felipecrs/repos/vscode-shellcheck/out/src/utils/tool-check.js
        at packageResolve (node:internal/modules/esm/resolve:880:9)
        at moduleResolve (node:internal/modules/esm/resolve:953:18)
        at defaultResolve (node:internal/modules/esm/resolve:1195:11)
        at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:642:12)
        at #cachedDefaultResolve (node:internal/modules/esm/loader:591:25)
        at ModuleLoader.getModuleJobForRequire (node:internal/modules/esm/loader:347:53)
        at new ModuleJobSync (node:internal/modules/esm/module_job:333:34)
        at ModuleLoader.getModuleJobForRequire (node:internal/modules/esm/loader:407:11)
        at new ModuleJobSync (node:internal/modules/esm/module_job:333:34)
        at ModuleLoader.importSyncForRequire (node:internal/modules/esm/loader:320:11)
        at loadESMFromCJS (node:internal/modules/cjs/loader:1378:24)
        at Module._compile (node:internal/modules/cjs/loader:1518:5)
        at Module._extensions..js (node:internal/modules/cjs/loader:1588:16)
        at Module.load (node:internal/modules/cjs/loader:1282:32)
        at Module._load (node:internal/modules/cjs/loader:1103:12)
        at c._load (node:electron/js2c/node_init:2:16955)
        at e._load (file:///home/felipecrs/repos/vscode-shellcheck/.vscode-test/vscode-linux-x64-1.100.0/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:225:1509)
        at t._load (file:///home/felipecrs/repos/vscode-shellcheck/.vscode-test/vscode-linux-x64-1.100.0/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:180:22628)
        at r._load (file:///home/felipecrs/repos/vscode-shellcheck/.vscode-test/vscode-linux-x64-1.100.0/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:172:22062)
        at Module.require (node:internal/modules/cjs/loader:1305:19)
        at require (node:internal/modules/helpers:182:18)
        at /home/felipecrs/repos/vscode-shellcheck/node_modules/mocha/lib/mocha.js:416:36
        at Array.forEach (<anonymous>)
        at Mocha.loadFiles (/home/felipecrs/repos/vscode-shellcheck/node_modules/mocha/lib/mocha.js:413:14)
        at Mocha.run (/home/felipecrs/repos/vscode-shellcheck/node_modules/mocha/lib/mocha.js:988:10)
        at mochaGlobalTeardown.mochaGlobalTeardown (/home/felipecrs/repos/vscode-shellcheck/node_modules/@vscode/test-cli/out/runner.cjs:33:50)
        at new Promise (<anonymous>)
        at Object.run (/home/felipecrs/repos/vscode-shellcheck/node_modules/@vscode/test-cli/out/runner.cjs:33:11)
        at file:///home/felipecrs/repos/vscode-shellcheck/.vscode-test/vscode-linux-x64-1.100.0/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:123:17846
        at new Promise (<anonymous>)
        at PJ.sb (file:///home/felipecrs/repos/vscode-shellcheck/.vscode-test/vscode-linux-x64-1.100.0/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:123:17567)
        at async PJ.$extensionTestsExecute (file:///home/felipecrs/repos/vscode-shellcheck/.vscode-test/vscode-linux-x64-1.100.0/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:123:17155)
[main 2025-05-10T20:13:22.099Z] Extension host with pid 87078 exited with code: 0, signal: unknown.
Exit code:   1
```

Has someone run into this issue too?

**Comment by s-h-a-d-o-w** [comment]
@felipecrs Based on the sample, we have to create and run this wrapper instead of `vscode-test` now: https://github.com/jrieken/vscode-esm-sample-extension/blob/main/test/runTest.mjs

**Comment by felipecrs** [comment]
> [@felipecrs](https://github.com/felipecrs) Based on the sample, we have to create and run this wrapper instead of `vscode-test` now: [jrieken/vscode-esm-sample-extension@`main`/test/runTest.mjs](https://github.com/jrieken/vscode-esm-sample-extension/blob/main/test/runTest.mjs?rgh-link-date=2025-05-17T10%3A56%3A27.000Z)

I think that wrapper was based on when `vscode-test` cli was not a thing yet. I remember having similar wrappers back then. I hope `vscode-test` cli gets support for ESM at some point.

And thanks @s-h-a-d-o-w, I'll look into using the wrapper for now.

**Comment by itsdouges** [comment]
Has anyone been able to run esm + typescript from source?

Previously you could use register hooks. I don't believe we have access to import hooks?

**Comment by felipecrs** [comment]
I reported the vscode-test-cli issue at:

- https://github.com/microsoft/vscode-test-cli/issues/77

Any love from vscode team would be awesome!

**Comment by felipecrs** [comment]
> I reported the vscode-test-cli issue at:
> 
> * [Support for ESM extensions vscode-test-cli#77](https://github.com/microsoft/vscode-test-cli/issues/77)
> 
> Any love from vscode team would be awesome!

That's the only missing piece for [migrating VS Code ShellCheck to ESM](https://github.com/vscode-shellcheck/vscode-shellcheck/pull/1614). 😞

**Comment by AlexCannonball** [comment]
Hello.

An extension migrated to ESM works fine with `mocha 11.6.0` as a test runner for `@vscode/test-electron`.

However when upgrading to `mocha 11.7.0` the test instance of VS Code opens but never starts a test run. I guess it's related to mochajs/mocha#5366. Should this issue be reported here or in the Mocha project?

**Comment by unional** [comment]
I'm getting `Dynamic require of "path" is not supported.` when trying to convert https://github.com/unional/vscode-sort-package-json to ESM.

Need to find out what's causing that.

**Comment by verhovsky** [comment]
ESM was introduced in 2015. This issue is the last barrier left to their complete adoption, at least for me.

**Comment by MicahZoltu** [comment]
Status of Web Worker support for import maps:
https://github.com/WICG/import-maps/issues/2
https://github.com/whatwg/html/pull/10858

The https://github.com/whatwg/html/pull/10858#issuecomment-3790005909 from the author of that PR suggests that this is stalled just because they don't want to push it over the finish line if no one is going to implement.  So it seems that at the moment someone needs to convince browsers to step up and actually care about finishing import-map support.

**Comment by lishaduck** [comment]
I've been following this issue for a while and as I was just cleaning up my house I was thinking about this. (About to start a new extension in a few days 🤞🏻)

Would it be possible to shim out web worker support here? https://github.com/guybedford/es-module-shims was the import map polyfill back in the day, and it supported web workers. I just took another look and judging from the issues, I think vs code could still use it to shim out the `vscode` import in workers so long as they're not service workers.

Apologies if a similar approach has already been considered, it just seemed from the comments above that the team was open to alternatives to waiting on standardization.

(Also, my apologies for any typos, I'm typing this on my phone because I wanted to get it down before I forgot)

EDIT: And now that I posted, this probably should have gone in the loader repository but oh well 🤷🏻‍♂️

**Comment by alexdima** [comment]
🤖 Thanks for the report and the follow-up context. This is now implemented for the Node extension host in commit 15260e72ae15426bf6093b1ad97e7f29bbc2397e via https://github.com/microsoft/vscode/pull/245214 and is contained in 1.100.0 and later. Current main detects .mjs / package type "module" extension entry points and loads them with import(...), and https://github.com/microsoft/vscode/pull/248286 added the hybrid path for extensions that use ESM in the Node extension host and CJS in the web worker extension host. I think this can be closed as completed; any remaining request for ESM in the web worker extension host should be tracked separately.

**Comment by brianjenkins94** [comment]
Does there exist an issue for ESM in the web worker extension host?
